README
This assignment comes with two parts that are both intertwined. The first part was an idea
I had when we were given the opprotunity to choose our own projects. I saw that there wasn't
much incentive to travel up and down stairs, as well as little to risk vs. reward in the game.
So, to alleviate those concerns, I created a new character, Dave. A group of explorers went
into the dungeon before you, a group of Daves that were seeking out a beautiful treasure in
the depths of the dungeon. They discovered the treasure, but the boss of the dungeon kidnapped
all of the Daves and implanted explosives within thier bodies. You must now rescue the Daves
from the dungeon by walking over them, adding them to your band of Daves. If you make it to
the stairs and ascend/descend the stairs, you will have freed the Daves from their prison,
and in their joy, they give you the gold that they acquired. The risk with recruiting Daves
is that if they die, they set off a massive explosion, damaging everyone in a 2 block radius.
The second portion of the assignment was to implement a store, which is extremely
straightforward. Using the rewards for getting Daves to safety, you can spend it at the store
to get items without picking them up off of the dungeon floor.

CHANGELOG
1/18/2018: Wrote assignment1.01
1/18/2018: Wrote Makefile
1/19/2018: Bugfixed infinite loop errors in assignment1.01
1/20/2018: Bugfixed array boundary erros in assignment1.01
1/20/2018: Wrote CHANGELOG and README
1/25/2018: Changed the dungeon struct to be a pointer
1/26/2018: Changed the room_t struct to dynamically allocate space
1/27/2018: Wrote the arguments and baseline for saving and loading dungeons
1/29/2018: Finished writing and bugfixing saving and loading dungeons (assignment1.02)
1/31/2018: Wrote priorityQueue.c and priorityQueue.h
2/01/2018: Wrote hero struct, weight_dungeon, full_distance_graph, rooms_distance_graph, and print_distance_graph
2/01/2018: Wrote dijkstra.c and dijkstra.h
2/08/2018: Began bugfixing dijkstra.c and priorityQueue.c
2/10/2018: Continued bugfixing dijkstra.c and priorityQueue.c
2/11/2018: Continued bugfixing dijkstra.c and priorityQueue.c
2/12/2018: Finished bugfixing dijkstra.c and priorityQueue.c (assignment1.03)
2/15/2018: Created monster.c
2/16/2018: Created character.c and hero.c
2/18/2018: Created move_character in character.c
2/19/2018: Created character.h, hero.h, monster.h, dungeon.h
2/20/2018: Bugfixed and finished assignment1.04
2/27/2018: Initialized screen and inputs for integration with ncurses
2/27/2018: Changed printf statements within print_dungeon() to printw statements
3/01/2018: Wrote keybinds for hero movement as well as staircase methods
3/01/2018: Created method to view monster list and exit without movement
3/04/2018: Worked on keybinds and methods to scroll the monster list
3/05/2018: Fixed issue where list wouldn't scroll and update
3/05/2018: Fixed issue where dungeon wouldn't build after --nummon was specified
3/05/2018: Finalized and submitted implementation of ncurses and staircases (assignment1.05)
3/17/2018: Moved to professor's code
3/17/2018: Converted .c files into .cpp files. Changed character, pc, and npc to classes 
3/18/2018: Added debug mode for fog of war and teleportation
3/19/2018: Bugfixing
3/20/2018: Finished bugfixing for assignment1.06
3/25/2018: Wrote parsing monster definitions
3/26/2018: Finished writing monster definitions and started bugfixing
3/27/2018: Finished and submitted parsing mosnter definitions
4/01/2018: Created object.cpp and object.h
4/01/2018: Modified character and npc to fit monster descriptions
4/01/2018: Created object and monster factories in descriptions.cpp
4/02/2018: Modified io.cpp to render objects and colors
4/02/2018: Started bugfixing
4/03/2018: Finished bugfixing for assignment1.08
4/07/2018: Changed pc.cpp to carry vectors of objects for inventory and equipment
4/07/2018: Added methods to pc.cpp to wear, drop, take off, pick up, delete objects
4/07/2018: Changed io.cpp to display and work with the new commands for inventory and equipment
4/08/2018: Started bugfixing
4/09/2018: Finished bugfixing for assignment1.09
4/15/2018: Started dave.h and dave.cpp
4/16/2018: Incorporated Dave into combat as well as a vector within PC to follow Dave
4/16/2018: Updated and finished combat, including explosion damage
4/17/2018: Added the rewards system to sucessfully rescuing Daves
4/17/2018: Implemented a store to the I/O that the PC can use their rewards on
4/18/2018: Finished store implementation
4/18/2018: Major bugfixes
4/19/2018: Finished minor bugfixes to both Dave and the store
4/19/2018: Balance changes to explosion damage vs. rewards

move.c
#include "move.h"

#include <unistd.h>
#include <stdlib.h>
#include <assert.h>

#include "dungeon.h"
#include "heap.h"
#include "move.h"
#include "npc.h"
#include "pc.h"
#include "character.h"
#include "utils.h"
#include "path.h"
#include "event.h"
#include "io.h"

void do_combat(dungeon_t *d, character_t *atk, character_t *def)
{
  int can_see_atk, can_see_def;
  char *organs[] = {
    "liver",                   /*  0 */
    "pancreas",                /*  1 */
    "heart",                   /*  2 */
    "eye",                     /*  3 */
    "arm",                     /*  4 */
    "leg",                     /*  5 */
    "intestines",              /*  6 */
    "gall bladder",            /*  7 */
    "lungs",                   /*  8 */
    "hand",                    /*  9 */
    "foot",                    /* 10 */
    "spinal cord",             /* 11 */
    "pituitary gland",         /* 12 */
    "thyroid",                 /* 13 */
    "tongue",                  /* 14 */
    "bladder",                 /* 15 */
    "diaphram",                /* 16 */
    "stomach",                 /* 17 */
    "pharynx",                 /* 18 */
    "esophagus",               /* 19 */
    "trachea",                 /* 20 */
    "urethra",                 /* 21 */
    "spleen",                  /* 22 */
    "ganglia",                 /* 23 */
    "ear",                     /* 24 */
    "subcutaneous tissue"      /* 25 */
    "cerebellum",              /* 26 */ /* Brain parts begin here */
    "hippocampus",             /* 27 */
    "frontal lobe",            /* 28 */
    "brain",                   /* 29 */
  };
  int part;

  if (def->alive) {
    def->alive = 0;
    charpair(def->position) = NULL;
    
    if (def != &d->pc) {
      d->num_monsters--;
    } else {
      if ((part = rand() % (sizeof (organs) / sizeof (organs[0]))) < 26) {
        io_queue_message("As the %c eats your %s, "
                         "you wonder if there is an afterlife.",
                         atk->symbol, organs[part]);
      } else {
        io_queue_message("Your last thoughts fade away as "
                         "the %c eats your %s...",
                         atk->symbol, organs[part]);
      }
      /* Queue an empty message, otherwise the game will not pause for *
       * player to see above.                                          */
      io_queue_message("");
    }
    atk->kills[kill_direct]++;
    atk->kills[kill_avenged] += (def->kills[kill_direct] +
                                  def->kills[kill_avenged]);
  }

  if (atk == &d->pc) {
    io_queue_message("You smite the %c!", def->symbol);
  }

  can_see_atk = can_see(d, &d->pc, atk);
  can_see_def = can_see(d, &d->pc, def);

  if (atk != &d->pc && def != &d->pc) {
    if (can_see_atk && !can_see_def) {
      io_queue_message("The %c callously murders some poor, "
                       "defenseless creature.", atk->symbol);
    }
    if (can_see_def && !can_see_atk) {
      io_queue_message("Something kills the helpless %c.", def->symbol);
    }
    if (can_see_atk && can_see_def) {
      io_queue_message("You watch in abject horror as the %c "
                       "gruesomely murders the %c!", atk->symbol, def->symbol);
    }
  }
}

void move_character(dungeon_t *d, character_t *c, pair_t next)
{
  if (charpair(next) &&
      ((next[dim_y] != c->position[dim_y]) ||
       (next[dim_x] != c->position[dim_x]))) {
    do_combat(d, c, charpair(next));
  } else {
    /* No character in new position. */

    d->character[c->position[dim_y]][c->position[dim_x]] = NULL;
    c->position[dim_y] = next[dim_y];
    c->position[dim_x] = next[dim_x];
    d->character[c->position[dim_y]][c->position[dim_x]] = c;
  }
}

void do_moves(dungeon_t *d)
{
  pair_t next;
  character_t *c;
  event_t *e;

  /* Remove the PC when it is PC turn.  Replace on next call.  This allows *
   * use to completely uninit the heap when generating a new level without *
   * worrying about deleting the PC.                                       */

  if (pc_is_alive(d)) {
    /* The PC always goes first one a tie, so we don't use new_event().  *
     * We generate one manually so that we can set the PC sequence       *
     * number to zero.                                                   */
    e = malloc(sizeof (*e));
    e->type = event_character_turn;
    /* Hack: New dungeons are marked.  Unmark and ensure PC goes at d->time, *
     * otherwise, monsters get a turn before the PC.                         */
    if (d->is_new) {
      d->is_new = 0;
      e->time = d->time;
    } else {
      e->time = d->time + (1000 / d->pc.speed);
    }
    e->sequence = 0;
    e->c = &d->pc;
    heap_insert(&d->events, e);
  }

  while (pc_is_alive(d) &&
         (e = heap_remove_min(&d->events)) &&
         ((e->type != event_character_turn) || (e->c != &d->pc))) {
    d->time = e->time;
    if (e->type == event_character_turn) {
      c = e->c;
    }
    if (!c->alive) {
      if (d->character[c->position[dim_y]][c->position[dim_x]] == c) {
        d->character[c->position[dim_y]][c->position[dim_x]] = NULL;
      }
      if (c != &d->pc) {
        event_delete(e);
      }
      continue;
    }

    npc_next_pos(d, c, next);
    move_character(d, c, next);

    heap_insert(&d->events, update_event(d, e, 1000 / c->speed));
  }

  io_display(d);
  if (pc_is_alive(d) && e->c == &d->pc) {
    c = e->c;
    d->time = e->time;
    /* Kind of kludgey, but because the PC is never in the queue when   *
     * we are outside of this function, the PC event has to get deleted *
     * and recreated every time we leave and re-enter this function.    */
    e->c = NULL;
    event_delete(e);
    io_handle_input(d);
  }
}

void dir_nearest_wall(dungeon_t *d, character_t *c, pair_t dir)
{
  dir[dim_x] = dir[dim_y] = 0;

  if (c->position[dim_x] != 1 && c->position[dim_x] != DUNGEON_X - 2) {
    dir[dim_x] = (c->position[dim_x] > DUNGEON_X - c->position[dim_x] ? 1 : -1);
  }
  if (c->position[dim_y] != 1 && c->position[dim_y] != DUNGEON_Y - 2) {
    dir[dim_y] = (c->position[dim_y] > DUNGEON_Y - c->position[dim_y] ? 1 : -1);
  }
}

uint32_t against_wall(dungeon_t *d, character_t *c)
{
  return ((mapxy(c->position[dim_x] - 1,
                 c->position[dim_y]    ) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x] + 1,
                 c->position[dim_y]    ) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x]    ,
                 c->position[dim_y] - 1) == ter_wall_immutable) ||
          (mapxy(c->position[dim_x]    ,
                 c->position[dim_y] + 1) == ter_wall_immutable));
}

uint32_t in_corner(dungeon_t *d, character_t *c)
{
  uint32_t num_immutable;

  num_immutable = 0;

  num_immutable += (mapxy(c->position[dim_x] - 1,
                          c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x] + 1,
                          c->position[dim_y]    ) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
                          c->position[dim_y] - 1) == ter_wall_immutable);
  num_immutable += (mapxy(c->position[dim_x]    ,
                          c->position[dim_y] + 1) == ter_wall_immutable);

  return num_immutable > 1;
}

static void new_dungeon_level(dungeon_t *d, uint32_t dir)
{
  /* Eventually up and down will be independantly meaningful. *
   * For now, simply generate a new dungeon.                  */

  switch (dir) {
  case '<':
  case '>':
    new_dungeon(d);
    break;
  default:
    break;
  }
}


uint32_t move_pc(dungeon_t *d, uint32_t dir)
{
  pair_t next;
  uint32_t was_stairs = 0;

  next[dim_y] = d->pc.position[dim_y];
  next[dim_x] = d->pc.position[dim_x];


  switch (dir) {
  case 1:
  case 2:
  case 3:
    next[dim_y]++;
    break;
  case 4:
  case 5:
  case 6:
    break;
  case 7:
  case 8:
  case 9:
    next[dim_y]--;
    break;
  }
  switch (dir) {
  case 1:
  case 4:
  case 7:
    next[dim_x]--;
    break;
  case 2:
  case 5:
  case 8:
    break;
  case 3:
  case 6:
  case 9:
    next[dim_x]++;
    break;
  case '<':
    if (mappair(d->pc.position) == ter_stairs_up) {
      was_stairs = 1;
      new_dungeon_level(d, '<');
    }
    break;
  case '>':
    if (mappair(d->pc.position) == ter_stairs_down) {
      was_stairs = 1;
      new_dungeon_level(d, '>');
    }
    break;
  }

  if (was_stairs) {
    return 0;
  }

  if ((dir != '>') && (dir != '<') && (mappair(next) >= ter_floor)) {
    move_character(d, &d->pc, next);
    dijkstra(d);
    dijkstra_tunnel(d);

    return 0;
  }

  return 1;
}
